### DISCUSSION TITLE
Extended benchmark

### DISCUSSION THREAD

**Discussion by mmikhasenko** [ComPWA/polarimetry#68] [General]
thoughts to come back after the sensitivity project!

The difference in the speed between SymPy+JIT and Julia still bothers me.
I have a guess that the most of it comes from taking the summation over the helicity indices analytically.

I would be interested to follow the polarization asymmetry project with this investigation.
Particularly, we would like to do an extended benchmark.
 - different decay chains have different complexity, the time will differ.
 - the unaligned chains do not have summation, can be checked.
 - implementation of wigner d and clabsches is to be benchmarked

There is a great interest in faster amplitude fitter in LHCb. The toys we do not for Lb decays will take a month on 10 GPU. A factor of 100, or even 10 would matter a lot.

**Comment by redeboer** [comment]
This week I saw some peculiar behaviour that makes me doubt the JAX benchmarks: it is slower when parameters update, which suggests some form of caching. I haven't noticed this before and IPython's [`%time`](https://ipython.readthedocs.io/en/stable/interactive/magics.html#magic-time) should warn if a second run is significantly faster, but it's definitely there when rerunning for instance the cell with the intensity matrix. I also noticed that the loop over the different models in the `statistics.ipynb` notebook has a small 'hick-up' when it iterates over the model that has a different $R$.

My first guess was that JAX v0.3.x introduced some new behaviour, but I don't see any specific new caching features in [their changelog](https://jax.readthedocs.io/en/latest/changelog.html). At any rate, has to be investigated. A first step may be to separate the 'benchmarks' into a separate notebook, so that no intensities are computed prior to the `%%timeit` cell magic.

**Reply by redeboer** [reply]
See https://github.com/redeboer/polarization-sensitivity/pull/69

**Comment by redeboer** [comment]
> * implementation of wigner d and clabsches is to be benchmarked

As a side note, I noticed that sympy does not simplify expressions like
```python
sp.sin(sp.acos((x+y)/z))
```
or at least certain combined expressions. This would also have to be investigated.

**Comment by mmikhasenko** [comment]
> As a side note, I noticed that sympy does not simplify expressions like
> 
> ```python
> sp.sin(sp.acos((x+y)/z))
> ```
> 
> or at least certain combined expressions. This would also have to be investigated.

yes, I worked around that by replacing the sin and cos separately.
I think it is ok. the expression for sin are even simpler than the cos.
